### PR TITLE
feat(db2): Add initial support

### PR DESCRIPTION
**Pull request by D3f0**
The motivation for this pull request is to add support for [IBM DB2](https://www.ibm.com/docs/en/db2-for-zos/12?topic=db2-sql) dialect to parse DDL and DML for DB2. At this moment *not all the syntax* is supported but should be sufficient for basic interaction.

Some of the code was also tested using the [Python testcontainers package](https://github.com/testcontainers/testcontainers-python/blob/main/modules/db2/testcontainers/db2/__init__.py) using DB2's Docker based community image: `icr.io/db2_community/db2:latest`.

Notes:
This is Farah Abdirahman's work adapted to sqlglot commit guidelines.

### CODE REVIEW

File: sqlglot/dialects/db2.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+    class Tokenizer(tokens.Tokenizer):
+        QUOTES = ["'", '"']
+        IDENTIFIERS = ["`"]
+
+        # Associates certain meaningful words with tokens that capture their intent
+        KEYWORDS = {
+            **Tokenizer.KEYWORDS,
+            "(+)": TokenType.JOIN_MARKER,
+            "INT64": TokenType.BIGINT,
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:91]: This is redundant, the base tokenizer already includes this in `KEYWORDS`.

File: sqlglot/dialects/db2.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            "FLOAT64": TokenType.DOUBLE,
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:92]: This is repeated across multiple dialects, let's move it to the base tokenizer's `KEYWORDS` and get rid of the other entries.

File: sqlglot/dialects/db2.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+            "CURRENT DATE": TokenType.CURRENT_DATE,
+            "CURRENT TIME": TokenType.CURRENT_TIME,
+            "CURRENT TIMESTAMP": TokenType.CURRENT_TIMESTAMP,
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:95]: These keywords look questionable, does db2 actually expect `CURRENT X`, i.e. separated using a space?

File: sqlglot/dialects/db2.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+            "ORDER BY": TokenType.ORDER_BY,
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:96]: Redundant, it's included in base `KEYWORDS`.

```suggestion
```

File: sqlglot/dialects/db2.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        }
+
+    class Parser(parser.Parser):
+        ALTER_TABLE_ADD_REQUIRED_FOR_EACH_COLUMN = False
+        WINDOW_BEFORE_PAREN_TOKENS = {TokenType.OVER, TokenType.KEEP}
+        VALUES_FOLLOWED_BY_PAREN = False
+        ALIAS_POST_TABLESAMPLE = True
+        STRING_ALIASES = True
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:104]: Let's make sure to test these.

File: sqlglot/dialects/db2.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            "DATE_SUB": build_date_delta_with_interval(exp.DateSub),
+            "TO_CHAR": _build_timetostr_or_tochar,
+            "DATE_ADD": build_date_delta_with_interval(exp.DateAdd),
+            "DATE": lambda args: exp.TsOrDsToDate(this=seq_get(args, 0)),
+            "CHAR": lambda self: self._parse_chr(),
+            "GROUP_CONCAT": lambda self: self._parse_group_concat(),
+            # https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_values
+            "VALUES": lambda self: self.expression(
+                exp.Anonymous, this="VALUES", expressions=[self._parse_id_var()]
+            ),
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:117]: Why are we including a link to MySQL's docs here?

File: sqlglot/dialects/db2.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+            "TO_DATE": build_formatted_time(exp.StrToDate, "db2"),
+            "DAY": lambda args: exp.Day(this=exp.TsOrDsToDate(this=seq_get(args, 0))),
+            "TO_DAYS": lambda args: exp.paren(
+                exp.DateDiff(
+                    this=exp.TsOrDsToDate(this=seq_get(args, 0)),
+                    expression=exp.TsOrDsToDate(this=exp.Literal.string("0000-01-01")),
+                    unit=exp.var("DAY"),
+                )
+                + 1
+            ),
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:127]: Factor this out in `dialects.py` and reuse it for both MySQL and DB2.

File: sqlglot/dialects/db2.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        }
+
+        NO_PAREN_FUNCTIONS = {
+            **parser.Parser.NO_PAREN_FUNCTIONS,
+            TokenType.CURRENT_DATE: exp.CurrentDate,
+            TokenType.CURRENT_TIME: exp.CurrentTime,
+            TokenType.CURRENT_TIMESTAMP: exp.CurrentTimestamp,
+        }
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:135]: These are redundant.
```suggestion
```

File: sqlglot/dialects/db2.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+            "JSON_ARRAYAGG": lambda self: self._parse_json_array(
+                exp.JSONArrayAgg,
+                this=self._parse_format_json(self._parse_bitwise()),
+                order=self._parse_order(),
+            ),
+            "JSON_OBJECT": lambda self: self.elf._parse_json_object(
+                exp.JSONObject,
+                this=self._parse_format_json(self._parse_bitwise()),
+            ),
+            "XMLTABLE": lambda self: self._parse_xml_table(),
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:151]: There's a lot of Oracle code duplication here as well.

File: sqlglot/dialects/db2.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+        }
+        DISTINCT_TOKENS = {TokenType.DISTINCT}
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:153]: This is redundant.

```suggestion
```

File: sqlglot/dialects/db2.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+
+        QUERY_MODIFIER_PARSERS = {
+            **parser.Parser.QUERY_MODIFIER_PARSERS,
+            TokenType.ORDER_BY: lambda self: ("order", self._parse_order()),
+        }
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:158]: This is redundant.

```suggestion
```

File: sqlglot/dialects/db2.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+
+    class Generator(generator.Generator):
+        # Specifies how AST nodes, i.e. subclasses of exp.Expression, should be converted into SQL
```
Comment (georgesittas) [inline sqlglot/dialects/db2.py:161]:
```suggestion
```

**Review by georgesittas — COMMENTED** [review]
Hey @D3f0 👋

Appreciate the PR, but there's quite a lot of cleaning to do here. Please check out the comments I left and try to reduce code duplication as much as possible. I'm curious whether `DB2` can actually inherit from `Oracle`, given the overlap between the two.

**Comment by georgesittas** [comment]
Another important thing: you need to document the motivation behind the implementation details for this to be accepted. We are not going to crawl DB2's in order to verify the assumptions here are correct, so please update the PR's description to help us with this.

File: sqlglot/dialects/db2.py [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+    this = seq_get(args, 0)
+
+    if this and not this.type:
+        from sqlglot.optimizer.annotate_types import annotate_types
+
+        annotate_types(this)
+        if this.is_type(*exp.DataType.TEMPORAL_TYPES):
+            return build_formatted_time(exp.TimeToStr, "db2", default=True)(args)
+
+    return exp.ToChar.from_arg_list(args)
```
[the remark this answers is not in the document]
Comment (D3f0) [reply sqlglot/dialects/db2.py:32]: @georgesittas should we remove the `_` prefix as well?

**Comment by D3f0** [comment]
> Another important thing: you need to document the motivation behind the implementation details for this to be accepted. We are not going to crawl DB2's docs in order to verify that the assumptions here are correct, so please update the PR's description accordingly.

I've updated the description of the pull request. Kindly let us know if more information is needed.

File: sqlglot/dialects/__init__.py [hunk 14]
```diff
@@ -87,3 +87,4 @@ class Generator(Generator):
 from sqlglot.dialects.teradata import Teradata
 from sqlglot.dialects.trino import Trino
 from sqlglot.dialects.tsql import TSQL
+from sqlglot.dialects.db2 import Db2
```
[the remark this answers is not in the document]
Comment (D3f0) [reply sqlglot/dialects/__init__.py]: Fixed in 1e96557b9b8456adb61c6cc1eb879902bf2ae9d6

**Comment by georgesittas** [comment]
Can you include links to the documentation that justify certain decisions in the dialect? E.g. the normalization strategy, the time mapping, certain flags in the parser/generator classes, the `CURRENT X` functions I commented on, etc. Something that will help us cross-check the dialect implementation.

File: sqlglot/dialects/db2.py (hunk 8 above)
Comment (D3f0) [reply sqlglot/dialects/db2.py]: Removed in 1e96557b9b8456adb61c6cc1eb879902bf2ae9d6

File: sqlglot/dialects/db2.py (hunk 12 above)
Comment (D3f0) [reply sqlglot/dialects/db2.py]: Comment removed in 1e96557b9b8456adb61c6cc1eb879902bf2ae9d6

**Comment by georgesittas** [comment]
Hey @D3f0, any plans to finish this?

**Pull request closed without merging**